kv/kvserver/rangelog: TestRangeLog failed

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14294808?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14294808?buildTab=artifacts#/) on master @ [559994e230ac3776dd1188e66678cce9205ab345](https://github.com/cockroachdb/cockroach/commits/559994e230ac3776dd1188e66678cce9205ab345):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog3142926049
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog3142926049
--- FAIL: TestRangeLog (37.97s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36500

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14308877?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14308877?buildTab=artifacts#/) on master @ [f84d66b9f750a85f96f671c74ab6c8ca71409a7b](https://github.com/cockroachdb/cockroach/commits/f84d66b9f750a85f96f671c74ab6c8ca71409a7b):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog3585172581
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog3585172581
--- FAIL: TestRangeLog (37.03s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14319265?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14319265?buildTab=artifacts#/) on master @ [41da9eccc216413c09cc66ad4c6bfa5f92c506e3](https://github.com/cockroachdb/cockroach/commits/41da9eccc216413c09cc66ad4c6bfa5f92c506e3):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog2524822685
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog2524822685
--- FAIL: TestRangeLog (37.12s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14330464?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14330464?buildTab=artifacts#/) on master @ [fcf0b5a1c95167ed00fa52c698d50980e83e2982](https://github.com/cockroachdb/cockroach/commits/fcf0b5a1c95167ed00fa52c698d50980e83e2982):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog1736586264
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog1736586264
--- FAIL: TestRangeLog (37.23s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14350054?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14350054?buildTab=artifacts#/) on master @ [f1257abf5a6b2684dc626d8c783ccf3c5bebc770](https://github.com/cockroachdb/cockroach/commits/f1257abf5a6b2684dc626d8c783ccf3c5bebc770):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog884240897
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog884240897
--- FAIL: TestRangeLog (28.00s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/rangelog.TestRangeLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14370559?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14370559?buildTab=artifacts#/) on master @ [5634a5ec3eba818febceae214678a399f22d7f0e](https://github.com/cockroachdb/cockroach/commits/5634a5ec3eba818febceae214678a399f22d7f0e):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog610438798
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/61f20d2036ad4a00e3051978f8294f77/logTestRangeLog610438798
--- FAIL: TestRangeLog (38.73s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/rangelog.TestRangeLog [failed](https://tanzanite.cluster.engflow.com/invocations/default/2b522dbe-bb94-48ce-bb40-11605d361bef?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIvcmFuZ2Vsb2c6cmFuZ2Vsb2dfdGVzdA==) on master @ [5e6335f597fb7d01f28a3e081094472818e80e2c](https://github.com/cockroachdb/cockroach/commits/5e6335f597fb7d01f28a3e081094472818e80e2c):


```
=== RUN   TestRangeLog
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRangeLog2746006357
    test_log_scope.go:81: use -show-logs to present logs inline
    rangelog_test.go:102: error executing '
        SELECT crdb_internal.pretty_key(key, 1), 
               substring(encode(val, 'hex') from 9) -- strip the checksum
          FROM crdb_internal.scan(crdb_internal.index_span($1, 1)) as t(key, val)': pq: context canceled
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRangeLog2746006357
--- FAIL: TestRangeLog (34.48s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -2701,7 +2701,14 @@ func (ds *DistSender) sendToReplicas(
 		} else {
 			br, err = transport.SendNext(sendCtx, requestToSend)
 			tEnd := timeutil.Now()
-			cbToken.Done(br, err, tEnd.UnixNano())
+			if cancelErr := cbToken.Done(br, err, tEnd.UnixNano()); cancelErr != nil {
+				// The request was cancelled by the circuit breaker tripping. If this is
+				// detected by request evaluation (as opposed to the transport send), it
+				// will return the context error in br.Error instead of err, which won't
+				// be retried below. Instead, record it as a send error in err and retry
+				// when possible. This commonly happens when the replica is local.
+				br, err = nil, cancelErr
+			}
 
 			if dur := tEnd.Sub(tBegin); dur > slowDistSenderReplicaThreshold {
 				var s redact.StringBuilder

--- a/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker.go
@@ -448,15 +448,28 @@ func newReplicaCircuitBreaker(
 // replicaCircuitBreakerToken carries request-scoped state between Track() and
 // done().
 type replicaCircuitBreakerToken struct {
-	r      *ReplicaCircuitBreaker // nil if circuit breakers were disabled
-	ctx    context.Context
-	ba     *kvpb.BatchRequest
-	cancel context.CancelCauseFunc
+	// r is the circuit breaker reference. nil if circuit breakers were disabled
+	// when we began tracking the request.
+	r *ReplicaCircuitBreaker
+
+	// ctx is the client's original context, to determine if it has gone away.
+	ctx context.Context
+
+	// cancelCtx is the child context used to cancel the request. nil if
+	// cancellation is disabled.
+	cancelCtx context.Context
+
+	// ba is the batch request being tracked.
+	ba *kvpb.BatchRequest
 }
 
-// Done records the result of the request and untracks it.
-func (t replicaCircuitBreakerToken) Done(br *kvpb.BatchResponse, err error, nowNanos int64) {
-	t.r.done(t.ctx, t.ba, br, err, nowNanos, t.cancel)
+// Done records the result of the request and untracks it. If the request was
+// cancelled by the circuit breaker, an appropriate context cancellation error
+// is returned.
+func (t replicaCircuitBreakerToken) Done(
+	br *kvpb.BatchResponse, sendErr error, nowNanos int64,
+) error {
+	return t.r.done(t.ctx, t.cancelCtx, t.ba, br, sendErr, nowNanos)
 }
 
 // id returns a string identifier for the replica.
@@ -585,10 +598,13 @@ func (r *ReplicaCircuitBreaker) Track(
 			CircuitBreakerProbeTimeout.Get(&r.d.settings.SV).Nanoseconds()
 
 		if !hasTimeout {
-			sendCtx, token.cancel = context.WithCancelCause(ctx)
+			var cancel context.CancelCauseFunc
+			sendCtx, cancel = context.WithCancelCause(ctx)
+			token.cancelCtx = sendCtx
+
 			reqKind := cbRequestKindFromBatch(ba)
 			r.mu.Lock()
-			r.mu.cancelFns[reqKind][ba] = token.cancel
+			r.mu.cancelFns[reqKind][ba] = cancel
 			r.mu.Unlock()
 		}
 	}
@@ -598,32 +614,50 @@ func (r *ReplicaCircuitBreaker) Track(
 
 // done records the result of a tracked request and untracks it. It is called
 // via replicaCircuitBreakerToken.Done().
+//
+// If the request was cancelled by the circuit breaker, an appropriate context
+// cancellation error is returned.
 func (r *ReplicaCircuitBreaker) done(
 	ctx context.Context,
+	cancelCtx context.Context,
 	ba *kvpb.BatchRequest,
 	br *kvpb.BatchResponse,
-	err error,
+	sendErr error,
 	nowNanos int64,
-	cancel context.CancelCauseFunc,
-) {
+) error {
 	if r == nil {
-		return // circuit breakers disabled when we began tracking the request
+		return nil // circuit breakers disabled when we began tracking the request
 	}
 
-	// Untrack the request, and clean up the cancel function.
+	// Untrack the request.
 	if inflightReqs := r.inflightReqs.Add(-1); inflightReqs < 0 {
 		log.Fatalf(ctx, "inflightReqs %d < 0", inflightReqs)
 	}
 
-	if cancel != nil {
+	// Detect if the circuit breaker cancelled the request, and prepare a
+	// cancellation error to return to the caller.
+	var cancelErr error
+	if cancelCtx != nil {
+		if sendErr != nil || br.Error != nil {
+			if cancelErr = cancelCtx.Err(); cancelErr != nil && ctx.Err() == nil { // check ctx last
+				log.VErrEventf(ctx, 2,
+					"request cancelled by tripped circuit breaker for %s: %s", r.id(), cancelErr)
+				cancelErr = errors.Wrapf(cancelErr, "%s is unavailable (circuit breaker tripped)", r.id())
+			}
+		}
+
+		// Clean up the cancel function.
 		reqKind := cbRequestKindFromBatch(ba)
 		r.mu.Lock()
+		cancel := r.mu.cancelFns[reqKind][ba]
 		delete(r.mu.cancelFns[reqKind], ba) // nolint:deferunlockcheck
 		r.mu.Unlock()
-		cancel(nil)
+		if cancel != nil {
+			cancel(nil)
+		}
 	}
 
-	// If this was a local send error, i.e. err != nil, we rely on RPC circuit
+	// If this was a local send error, i.e. sendErr != nil, we rely on RPC circuit
 	// breakers to fail fast. There is no need for us to launch a probe as well.
 	// This includes the case where either the remote or local node has been
 	// decommissioned.
@@ -634,8 +668,8 @@ func (r *ReplicaCircuitBreaker) done(
 	// can't know if this was because of a client timeout or not, so we assume
 	// there may be a problem with the replica. We will typically see recent
 	// successful responses too if that isn't the case.
-	if err != nil && ctx.Err() == nil {
-		return
+	if sendErr != nil && ctx.Err() == nil {
+		return cancelErr
 	}
 
 	// If we got a response from the replica (even a br.Error), it isn't stalled.
@@ -644,13 +678,14 @@ func (r *ReplicaCircuitBreaker) done(
 	//
 	// NB: we don't reset this to 0 when inflightReqs==0 to avoid unnecessary
 	// synchronization.
-	if err == nil {
+	if sendErr == nil {
 		r.stallSince.Store(nowNanos)
 	}
 
-	// Handle error responses. To record the response as an error, err is set
-	// non-nil. Otherwise, the response is recorded as a success.
-	if err == nil && br.Error != nil {
+	// Record error responses, by setting err non-nil. Otherwise, the response is
+	// recorded as a success.
+	err := sendErr
+	if sendErr == nil && br.Error != nil {
 		switch tErr := br.Error.GetDetail().(type) {
 		case *kvpb.NotLeaseHolderError:
 			// Consider NLHE a success if it contains a lease record, as the replica
@@ -681,7 +716,6 @@ func (r *ReplicaCircuitBreaker) done(
 		}
 	}
 
-	// Track errors.
 	if err == nil {
 		// On success, reset the error tracking.
 		r.errorSince.Store(0)
@@ -693,6 +727,9 @@ func (r *ReplicaCircuitBreaker) done(
 		// The replica has been failing for the past probe threshold, probe it.
 		r.breaker.Probe()
 	}
+
+	// Return the client cancellation error (if any).
+	return cancelErr
 }
 
 // launchProbe spawns an async replica probe that sends LeaseInfo requests to

--- a/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go
@@ -277,7 +277,7 @@ func benchmarkCircuitBreakersTrack(
 					assert.NoError(b, err)
 					return
 				}
-				cbToken.Done(br, sendErr, nowNanos)
+				_ = cbToken.Done(br, sendErr, nowNanos) // ignore cancellation error
 			}
 		}()
 	}

